### PR TITLE
[SYCL] Handle address space casts in the LowerWGScope

### PR DESCRIPTION
**Pull request by againull**

### CODE REVIEW

File: llvm/lib/SYCLLowerIR/LowerWGScope.cpp [hunk 1]
```diff
@@ -779,13 +786,15 @@ PreservedAnalyses SYCLLowerWGScopePass::run(Function &F, const llvm::Triple &TT,
     // globals.
     Instruction *I = BB.getFirstNonPHI();
 
-    for (; I->getOpcode() == Instruction::Alloca; I = I->getNextNode()) {
+    for (; I->getOpcode() == Instruction::Alloca ||
```
Comment (DenisBakhvalov) [inline llvm/lib/SYCLLowerIR/LowerWGScope.cpp:789]: Maybe update the comment above since now you also skip addrspacecasts associated with allocas.

**Comment by DenisBakhvalov** [comment]
LGTM, however, please address CI failures.

**Comment by bader** [comment]
> LGTM, however, please address CI failures.

@againull, please, check if https://github.com/intel/llvm/pull/2864/commits/2b2ded9e5dc23603fa8729e20afdca7bd87c019d address the CI failure correctly.

File: llvm/lib/SYCLLowerIR/LowerWGScope.cpp (hunk 1 above)
Comment (againull) [reply llvm/lib/SYCLLowerIR/LowerWGScope.cpp:789]: Yes, thanks, fixed the comment.

**Comment by againull** [comment]
> > LGTM, however, please address CI failures.
> 
> @againull, please, check if [2b2ded9](https://github.com/intel/llvm/commit/2b2ded9e5dc23603fa8729e20afdca7bd87c019d) address the CI failure correctly.

Thanks, this commit is fixing the fail.

File: llvm/test/SYCLLowerIR/addrspacecast_handling.ll [hunk 2]
```diff
@@ -0,0 +1,83 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
+; RUN: opt < %s -LowerWGScope -S | FileCheck %s
+
```
Comment (kbobrovs) [inline llvm/test/SYCLLowerIR/addrspacecast_handling.ll:3]: Please add a comment what this test checks.

File: llvm/test/SYCLLowerIR/addrspacecast_handling.ll [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+; CHECK-NEXT:    call void @_Z22__spirv_ControlBarrierjjj(i32 2, i32 2, i32 272) [[ATTR0]]
+; CHECK-NEXT:    [[MAT_LD1:%.*]] = load [[STRUCT_HAM]] addrspace(4)*, [[STRUCT_HAM]] addrspace(4)* addrspace(3)* @[[SHADOW4]], align 8
+; CHECK-NEXT:    store [[STRUCT_HAM]] addrspace(4)* [[MAT_LD1]], [[STRUCT_HAM]] addrspace(4)** [[TMP]], align 8
+; CHECK-NEXT:    [[TMP8:%.*]] = bitcast %struct.spam* [[TMP3]] to i8*
+; CHECK-NEXT:    call void @llvm.memcpy.p0i8.p3i8.i64(i8* align 8 [[TMP8]], i8 addrspace(3)* align 16 bitcast (%struct.spam addrspace(3)* @[[SHADOW3]] to i8 addrspace(3)*), i64 36, i1 false)
+; CHECK-NEXT:    call void @_Z22__spirv_ControlBarrierjjj(i32 2, i32 2, i32 272) [[ATTR0]]
+; CHECK-NEXT:    [[TMP5:%.*]] = addrspacecast %struct.bar* [[ARG1]] to [[STRUCT_BAR:%.*]] addrspace(4)*
+; CHECK-NEXT:    [[TMP6:%.*]] = addrspacecast [[STRUCT_SPAM]] addrspace(4)* [[TMP4]] to %struct.spam*
+; CHECK-NEXT:    call spir_func void @widget(%struct.bar addrspace(4)* dereferenceable_or_null(32) [[TMP5]], %struct.spam* byval(%struct.spam) align 8 [[TMP6]])
+; CHECK-NEXT:    ret void
```
Comment (kbobrovs) [inline llvm/test/SYCLLowerIR/addrspacecast_handling.ll:64]: Nit:
Can the number of checks be reduced to addrspace cast dependence chain only?

**Comment by bader** [comment]
> > > LGTM, however, please address CI failures.
> > 
> > 
> > @againull, please, check if [2b2ded9](https://github.com/intel/llvm/commit/2b2ded9e5dc23603fa8729e20afdca7bd87c019d) address the CI failure correctly.
> 
> Thanks, this commit is fixing the fail.

I know, but I'd like you confirm that the fix is correct. :)

File: llvm/test/SYCLLowerIR/addrspacecast_handling.ll (hunk 3 above)
Comment (bader) [reply llvm/test/SYCLLowerIR/addrspacecast_handling.ll:64]: I think it's important to validate all the code generated by the pass to ensure correctness.
These checks should be easy to maintain as they are auto-generated (see the note at line 1).

File: llvm/lib/SYCLLowerIR/LowerWGScope.cpp [hunk 4]
```diff
@@ -630,6 +632,11 @@ static void fixupPrivateMemoryPFWILambdaCaptures(CallInst *PFWICall) {
   // whether it is an alloca with "work_item_scope"
   SmallVector<CaptureDesc, 4> PrivMemCaptures;
 
+  // Look through cast
+  auto *Cast = dyn_cast<AddrSpaceCastInst>(LambdaObj);
+  if (Cast)
```
Comment (bader) [inline llvm/lib/SYCLLowerIR/LowerWGScope.cpp:637]: From https://github.com/intel/llvm/pull/2864/files#r567811589:
Combine into `if (auto *Cast = dyn_cast<AddrSpaceCastInst>(LambdaObj))`

**Comment by againull** [comment]
> > > > LGTM, however, please address CI failures.
> > > 
> > > 
> > > @againull, please, check if [2b2ded9](https://github.com/intel/llvm/commit/2b2ded9e5dc23603fa8729e20afdca7bd87c019d) address the CI failure correctly.
> > 
> > 
> > Thanks, this commit is fixing the fail.
> 
> I know, but I'd like you confirm that the fix is correct. :)
> 
> UPDATE: I just noticed that checks in the test are auto-generated, so the right way to fix it is to run the script again. I made my changes manually.

Thanks, fixed.

File: llvm/lib/SYCLLowerIR/LowerWGScope.cpp (hunk 4 above)
Comment (againull) [reply llvm/lib/SYCLLowerIR/LowerWGScope.cpp]: Fixed.

File: llvm/test/SYCLLowerIR/addrspacecast_handling.ll (hunk 2 above)
Comment (againull) [reply llvm/test/SYCLLowerIR/addrspacecast_handling.ll:3]: Added comment.

File: llvm/test/SYCLLowerIR/addrspacecast_handling.ll (hunk 3 above)
Comment (againull) [reply llvm/test/SYCLLowerIR/addrspacecast_handling.ll:68]: I agree with Alexey here.

Comment (kbobrovs) [reply llvm/test/SYCLLowerIR/addrspacecast_handling.ll:68]: > I think it's important to validate all the code generated by the pass to ensure correctness.

I think regression tests are written for particular fixes and should be focused on testing the fix, not entire pass/feature, as there should exist other tests. But anyway, that was a nit, so either way is fine with me.
The script for automated maintenance is a good argument too.

**Review by kbobrovs — APPROVED** [review]

**Pull request merged**